src/climate_learn/data/climate_dataset/era5_module.py

```diff
@@ -1,94 +1,235 @@
-import os
+# Standard library
+import copy
 import glob
-import xarray as xr
-import numpy as np
+import os
+import random
+from typing import Callable, Dict, Iterable, Sequence, Tuple
 
+# Third party
+import numpy
+import torch
 from tqdm import tqdm
-from typing import Callable, Dict, Iterable, Sequence
-from climate_learn.data.climate_dataset import ClimateDataset
+import xarray as xr
+
+# Local application
 from climate_learn.data.climate_dataset.args import ERA5Args
-from ..constants import (
-    NAME_TO_VAR,
-    DEFAULT_PRESSURE_LEVELS,
+from climate_learn.data.climate_dataset import ClimateDataset
+from climate_learn.data.constants import (
     CONSTANTS,
-    SINGLE_LEVEL_VARS,
+    DEFAULT_PRESSURE_LEVELS,
+    NAME_LEVEL_TO_VAR_LEVEL,
+    NAME_TO_VAR,
     PRESSURE_LEVEL_VARS,
+    SINGLE_LEVEL_VARS,
 )
 
 
 class ERA5(ClimateDataset):
-    args_class: Callable[..., ERA5Args] = ERA5Args
+    _args_class: Callable[..., ERA5Args] = ERA5Args
 
     def __init__(self, data_args: ERA5Args) -> None:
         super().__init__(data_args)
         self.root_dir: str = data_args.root_dir
         self.years: Iterable[int] = data_args.years
 
-    def setup(self) -> None:
-        self.constant_names: Sequence[str] = []
-        self.data_dict: Dict[str, xr.core.dataarray.DataArray] = self.load_from_nc(
-            self.root_dir
+    def get_file_name_from_variable(self, var: str) -> str:
+        if var in SINGLE_LEVEL_VARS or var in PRESSURE_LEVEL_VARS:
+            return var
+        else:
+            return "_".join(var.split("_")[:-1])
+
+    def set_lat_lon(self, year: int) -> None:
+        # lat lon is stored in each of the nc files, just need to load one and extract
+        dir_var = os.path.join(
+            self.root_dir, self.get_file_name_from_variable(self.variables[0])
         )
-        self.set_lat_lon()
+        ps = glob.glob(os.path.join(dir_var, f"*{year}*.nc"))
+        xr_data = xr.open_mfdataset(ps, combine="by_coords")
+        self.lat: numpy.ndarray = xr_data["lat"].values
```

Comment:
We can do something like what you suggested but I don't want to return illegal values and then let the logic of calling class to handle whether the value is illegal or not.

Also, aside from this particular file. The `ModelModule` casually uses attributes from the `DataModule` such as `lat` and `lon`. I didn't wanted to go into the rabbit hole of changing the `ModelModule`. Thus you might see `ShardDataset` and `MapDataset` having `lat` and `lon` as attribute which is dataset specific. 

In the long run, I would like to have a function called `get_metadata()` and then it would become responsibility of `ModelModule` to take care of unpacking.

Reply:
Good points. I guess we have a lot of work to do to fully disentangle the model and data modules. 😅

Reply:
> We can do something like what you suggested but I don't want to return illegal values and then let the logic of calling class to handle whether the value is illegal or not.

How is this different from not having the parameter at all? For example, consider the following scenarios.

1. We do the current setup where `self.lat` is declared in `set_lat_lon`. Then, in the calling class, we need to do some check like `if hasattr(era5_module, "lat")`.
2. We do my proposed change. Then, in the calling class, we need to do some check like `if era5_module.lat`.

Either way, we have to make a check, right?

Reply:
We can do the second approach.

Reply:
I have gone ahead with what you suggested and initialized all the variables in the `__init__()`. Now, calling the `getter` functions for the attributes that has not been yet set would result in `None`.  It would then become responsibility of calling class which is `MapDataset` / `ShardDataset` in this case to deal with such illegal values.